Sources/Jobs/Scheduler/RecurrenceRule.swift

```diff
@@ -0,0 +1,129 @@
+        self.timeZone = timeZone
+    }
+
+    init(yearConstraint: YearRecurrenceRuleConstraint? = nil,
+         monthConstraint: MonthRecurrenceRuleConstraint? = nil,
+         dayOfMonthConstraint: DayOfMonthRecurrenceRuleConstraint? = nil,
+         dayOfWeekConstraint: DayOfWeekRecurrenceRuleConstraint? = nil,
+         hourConstraint: HourRecurrenceRuleConstraint? = nil,
+         minuteConstraint: MinuteRecurrenceRuleConstraint? = nil,
+         secondConstraint: SecondRecurrenceRuleConstraint? = nil,
+         timeZone: TimeZone = TimeZone.current) throws {
+        self.timeZone = timeZone
+        self.yearConstraint = yearConstraint
+        self.monthConstraint = monthConstraint
+        self.dayOfMonthConstraint = dayOfMonthConstraint
+        self.dayOfWeekConstraint = dayOfWeekConstraint
+        self.hourConstraint = hourConstraint
+        self.minuteConstraint = minuteConstraint
+        self.secondConstraint = secondConstraint
+    }
+
+    public mutating func setYearConstraint(_ yearConstraint: YearRecurrenceRuleConstraint) {
+        self.yearConstraint = yearConstraint
+    }
+
+    public mutating func setQuarterConstraint(_ quarterConstraint: QuarterRecurrenceRuleConstraint) {
+        self.quarterConstraint = quarterConstraint
+    }
+
+    public mutating func setMonthConstraint(_ monthConstraint: MonthRecurrenceRuleConstraint) {
+        self.monthConstraint = monthConstraint
+    }
+
+    public mutating func setDayOfMonthConstraint(_ dayOfMonthConstraint: DayOfMonthRecurrenceRuleConstraint) {
+        self.dayOfMonthConstraint = dayOfMonthConstraint
+    }
+
+    public mutating func setDayOfWeekConstraint(_ dayOfWeekConstraint: DayOfWeekRecurrenceRuleConstraint) {
+        self.dayOfWeekConstraint = dayOfWeekConstraint
+    }
+
+    public mutating func setHourConstraint(_ hourConstraint: HourRecurrenceRuleConstraint) {
+        self.hourConstraint = hourConstraint
+    }
+
+    public mutating func setMinuteConstraint(_ minuteConstraint: MinuteRecurrenceRuleConstraint) {
+        self.minuteConstraint = minuteConstraint
+    }
+
+    public mutating func setSecondConstraint(_ secondConstraint: SecondRecurrenceRuleConstraint) {
+        self.secondConstraint = secondConstraint
+    }
+
+    ///  Sets the timeZone used by rule constraintss
+    ///
+    /// - Parameter timeZone: The TimeZone constraints reference against
+    public mutating func usingTimeZone(_ timeZone: TimeZone) {
+        self.timeZone = timeZone
+    }
+}
+
+// RecurrenceRule Evaluation
+extension RecurrenceRule {
+    /// Finds the next date from the starting date that satisfies the rule
+    ///
+    /// - Warning: The search is exhausted after the year 3000
+    ///
+    /// - Parameter date: The starting date
+    /// - Returns: The next date that satisfies the rule
+    public func resolveNextDateThatSatisfiesRule(date: Date) throws -> Date {
+        fatalError("Not Implemented: resolveNextDateThatSatisfiesRule")
```

Comment:
Why isn't this implemented? Is it for the next PR? If so, let's remove for now to avoid `fatalError` making it into master

Reply:
This is where this pull request ends and the next PR will begin. This is the function that getting the next date that satisfies the rule given a starting date.

Reply:
Perfect - let's leave it for now, then